run/order.go

```diff
@@ -70,46 +77,40 @@ func Sort(root *config.Root, stacks config.List[*config.SortableStack]) (config.
 		)
 
 		if err != nil {
-			return nil, "", err
+			return "", err
 		}
 	}
 
 	reason, err := d.Validate()
 	if err != nil {
-		return nil, reason, err
+		return reason, err
 	}
 
 	order := d.Order()
-
-	orderedStacks := make(config.List[*config.SortableStack], 0, len(order))
-
-	isSelectedStack := func(s *config.Stack) bool {
-		// Stacks may be added on the DAG from after/before references
-		// but they should not be on the final order if they are not part
-		// of the previously selected stacks passed as a parameter.
-		// This is important for change detection to work on ordering and
-		// also for filtering by working dir.
-		for _, stack := range stacks {
-			if s.Dir == stack.Dir() {
-				return true
-			}
-		}
-		return false
-	}
-
-	for _, id := range order {
+	orderLookup := make(map[string]int, len(order))
+	for idx, id := range order {
 		val, err := d.Node(id)
 		if err != nil {
-			return nil, "", fmt.Errorf("calculating run-order: %w", err)
+			return "", fmt.Errorf("calculating run-order: %w", err)
 		}
 		s := val.(*config.Stack)
-		if !isSelectedStack(s) {
```

Comment:
can you explain why this can be removed now? because `BuildDAG()` can pull more stacks into the pool, then if we don't de-select the ones out of scope, the returned order will be wrong.

Reply:
As you said, before, we used a given list of stacks to construct a DAG (which may pull in more stacks), which gives us the ordering, then we de-select those that were not in the original list.

What we do now is using the ordering the DAG gives to map each entry of the original list to an ordering index. Then we use this index for sorting the original list.

Before, we were iterating `order`, which may contain more stacks than the original list, but now we just sort the original list, so nothing has to be removed there. The `orderLookup` map may contain entries for stacks that have been pulled in, but they don't affect the sorted list.

Reply:
ahhh yes now we do sort in-place in the provided items. Gotcha. Thanks for the explanation!